smartsim/_core/entrypoints/dragon.py

```diff
@@ -0,0 +1,178 @@
+import textwrap
+import typing as t
+from types import FrameType
+
+import zmq
+
+import smartsim._core.utils.helpers as _helpers
+from smartsim._core.launcher.dragon.dragonBackend import DragonBackend
+from smartsim._core.schemas import (
+    DragonBootstrapRequest,
+    DragonBootstrapResponse,
+    DragonShutdownResponse,
+)
+from smartsim._core.schemas.dragonRequests import request_serializer
+from smartsim._core.schemas.dragonResponses import response_serializer
+from smartsim._core.utils.network import get_best_interface_and_address
+
+# kill is not catchable
+SIGNALS = [signal.SIGINT, signal.SIGQUIT, signal.SIGTERM, signal.SIGABRT]
+
+SHUTDOWN_INITIATED = False
+
+
+def handle_signal(signo: int, _frame: t.Optional[FrameType]) -> None:
+    if not signo:
+        print("Received signal with no signo")
+    else:
+        print(f"Received {signo}")
+    cleanup()
+
+
+context = zmq.Context()
+
+"""
+Dragon server entrypoint script
+"""
+
+DBPID: t.Optional[int] = None
+
+
+def print_summary(network_interface: str, ip_address: str) -> None:
+    zmq_config = {"interface": network_interface, "address": ip_address}
+
+    with open("dragon_config.log", "w", encoding="utf-8") as dragon_config_log:
+        dragon_config_log.write(
+            textwrap.dedent(f"""\
+                -------- Dragon Configuration --------
+                IPADDRESS: {ip_address}
+                NETWORK: {network_interface}
+                HOSTNAME: {socket.gethostname()}
+                DRAGON_SERVER_CONFIG: {json.dumps(zmq_config)}
+                --------------------------------------
+
+                --------------- Output ---------------
+
+                """),
+        )
+
+
+def run(dragon_head_address: str) -> None:
+    global SHUTDOWN_INITIATED  # pylint: disable=global-statement
+    print(f"Opening socket {dragon_head_address}")
+    dragon_head_socket = context.socket(zmq.REP)
+    dragon_head_socket.bind(dragon_head_address)
+    dragon_backend = DragonBackend()
+
+    while not SHUTDOWN_INITIATED:
+        print(f"Listening to {dragon_head_address}")
+        req = dragon_head_socket.recv_json()
+        print(f"Received request: {req}")
+        drg_req = request_serializer.deserialize_from_json(str(req))
+        resp = dragon_backend.process_request(drg_req)
+        print(f"Sending response {resp}", flush=True)
+        dragon_head_socket.send_json(response_serializer.serialize_to_json(resp))
+        if isinstance(resp, DragonShutdownResponse):
+            SHUTDOWN_INITIATED = True
+
+
+def main(args: argparse.Namespace) -> int:
+    interface, address = get_best_interface_and_address()
+    if not interface:
+        raise ValueError("Net interface could not be determined")
+    dragon_head_address = f"tcp://{address}"
+
+    if args.launching_address:
+        if str(args.launching_address).split(":", maxsplit=1)[0] == dragon_head_address:
+            address = "localhost"
+            dragon_head_address = "tcp://localhost:5555"
```

Comment:
consider reducing nested/repetitive string-building and pulling out single parts being computed. it could make it easier to find the one place where the string is put together and ensure they're consistent.

```py
def main()...
  ...
  transport = "tcp://"
  port = 5555
  if dragon_head_address not in args.launching_address:
    address = args.launching_address
  if port not in args.launching_port:  # example of another component, not really here
    port = args.launching_port

  dragon_head_address = f"{transport}{address}:{port}"
    
```